### ISSUE TITLE
Raster tool, XYZ-tiles creation (MBTiles)

### ISSUE THREAD

**Issue by AMI-Systems** [qgis/QGIS#48491]
### What is the bug or the crash?

During creation of an MBTiles-file from OSM, WikiMediaMaps or other the program breaks at the level 14:

Parameter
Map: OSM
Minimum zoom 2
Max zoom 15

QGIS-Version: 3.24.2-Tisler
QGIS-Codeversion: 13c1a028
Qt-Version: 5.15.2
Python-Version: 3.9.5
GDAL-Version: 3.4.2
GEOS-Version: 3.10.2-CAPI-1.16.0
PROJ-Version: Rel. 9.0.0, March 1st, 2022
PDAL-Version: 2.3.0 (git-version: 1c17a7)
Algorithmus gestartet um: 2022-05-09T22:36:08
Algorithmus XYZ-Kacheln erzeugen (MBTiles) startet…
Eingabeparameter:
{ 'BACKGROUND_COLOR' : QColor(0, 0, 0, 0), 'DPI' : 96, 'EXTENT' : '9819890.000000000,10087000.000000000,6086300.000000000,6382540.000000000 [EPSG:3857]', 'METATILESIZE' : 4, 'OUTPUT_FILE' : 'D:/Dokumente/QGis/J/WikiMap_1-15.mbtiles', 'QUALITY' : 75, 'TILE_FORMAT' : 0, 'ZOOM_MAX' : 15, 'ZOOM_MIN' : 2 }

Using 8 CPU Threads:
Generating tiles for zoom level: 2
Generating tiles for zoom level: 3
Generating tiles for zoom level: 4
Generating tiles for zoom level: 5
Generating tiles for zoom level: 6
Generating tiles for zoom level: 7
Generating tiles for zoom level: 8
Generating tiles for zoom level: 9
Generating tiles for zoom level: 10
Generating tiles for zoom level: 11
Generating tiles for zoom level: 12
Generating tiles for zoom level: 13
Generating tiles for zoom level: 14
Traceback (most recent call last):
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 455, in processAlgorithm
self.generate(writer, parameters, context, feedback)
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 316, in generate
for result in threadPool.map(self.renderSingleMetatile, metatiles_by_zoom[zoom]):
File "C:\PROGRA~1\QGIS32~1.2\apps\Python39\lib\concurrent\futures\_base.py", line 608, in result_iterator
yield fs.pop().result()
File "C:\PROGRA~1\QGIS32~1.2\apps\Python39\lib\concurrent\futures\_base.py", line 445, in result
return self.__get_result()
File "C:\PROGRA~1\QGIS32~1.2\apps\Python39\lib\concurrent\futures\_base.py", line 390, in __get_result
raise self._exception
File "C:\PROGRA~1\QGIS32~1.2\apps\Python39\lib\concurrent\futures\thread.py", line 52, in run
result = self.fn(*self.args, **self.kwargs)
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 230, in renderSingleMetatile
self.writer.write_tile(tile, tileImage)
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 401, in write_tile
self._init_zoom_layer(tile.z)
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 393, in _init_zoom_layer
self._execute_sqlite("UPDATE metadata SET value='{}' WHERE name='bounds'".format(bounds))
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 373, in _execute_sqlite
conn.execute(cmd)
sqlite3.OperationalError: database is locked

Execution failed after 104.40 Sekunden (1 Minute 44 Sekunden)

Lade Ergebnis Layer
Algorithmus 'XYZ-Kacheln erzeugen (MBTiles)' beendet

### Steps to reproduce the issue

try to create an mbtile from OSM from zoom level 2 up to 4

### Versions

3.24.2-Tisler

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by lbartoletti** [comment]
It seems to be a concurrency problem.

Is it possible for you to test with this fix:

replace the method `_execute_sqlite` at line 370 in `python/plugins\processing\algs\qgis\TilesXYZ.py` for you: `C:\PROGRA1/QGIS321.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py`

by

```python
    def _execute_sqlite(self, *commands):
        with sqlite3.connect(self.filename) as conn:
            for cmd in commands:
                c = conn.cursor()
                c.execute(cmd)
                conn.commit()
```

**Comment by AMI-Systems** [comment]
no, it does not help. I got the error again:

QGIS-Version: 3.24.2-Tisler
QGIS-Codeversion: 13c1a028
Qt-Version: 5.15.2
Python-Version: 3.9.5
GDAL-Version: 3.4.2
GEOS-Version: 3.10.2-CAPI-1.16.0
PROJ-Version: Rel. 9.0.0, March 1st, 2022
PDAL-Version: 2.3.0 (git-version: 1c17a7)
Algorithmus gestartet um: 2022-05-19T18:47:26
Algorithmus XYZ-Kacheln erzeugen (MBTiles) startet…
Eingabeparameter:
{ 'BACKGROUND_COLOR' : QColor(0, 0, 0, 0), 'DPI' : 96, 'EXTENT' : '4832560.000000000,5280910.000000000,6000260.000000000,6331500.000000000 [EPSG:3857]', 'METATILESIZE' : 4, 'OUTPUT_FILE' : 'D:/Dokumente/QGis/QField-Sync/out-test.mbtiles', 'QUALITY' : 75, 'TILE_FORMAT' : 0, 'ZOOM_MAX' : 15, 'ZOOM_MIN' : 12 }

Using 8 CPU Threads:
Generating tiles for zoom level: 12
Generating tiles for zoom level: 13
Generating tiles for zoom level: 14
Traceback (most recent call last):
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 456, in processAlgorithm
self.generate(writer, parameters, context, feedback)
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 316, in generate
for result in threadPool.map(self.renderSingleMetatile, metatiles_by_zoom[zoom]):
File "C:\PROGRA~1\QGIS32~1.2\apps\Python39\lib\concurrent\futures\_base.py", line 608, in result_iterator
yield fs.pop().result()
File "C:\PROGRA~1\QGIS32~1.2\apps\Python39\lib\concurrent\futures\_base.py", line 445, in result
return self.__get_result()
File "C:\PROGRA~1\QGIS32~1.2\apps\Python39\lib\concurrent\futures\_base.py", line 390, in __get_result
raise self._exception
File "C:\PROGRA~1\QGIS32~1.2\apps\Python39\lib\concurrent\futures\thread.py", line 52, in run
result = self.fn(*self.args, **self.kwargs)
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 230, in renderSingleMetatile
self.writer.write_tile(tile, tileImage)
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 402, in write_tile
self._init_zoom_layer(tile.z)
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 394, in _init_zoom_layer
self._execute_sqlite("UPDATE metadata SET value='{}' WHERE name='bounds'".format(bounds))
File "C:\PROGRA~1/QGIS32~1.2/apps/qgis/./python/plugins\processing\algs\qgis\TilesXYZ.py", line 374, in _execute_sqlite
c.execute(cmd)
sqlite3.OperationalError: database is locked

Execution failed after 103.63 Sekunden (1 Minute 44 Sekunden)

Lade Ergebnis Layer
Algorithmus 'XYZ-Kacheln erzeugen (MBTiles)' beendet

**Comment by alexbruy** [comment]
Fixed by #54321.
